R/super.R

```diff
@@ -36,20 +36,80 @@
 super <- function(...) {
   # TODO: (RK) Error on anonymous function calls.
   fn <- as.character(sys.call(-1)[[1]])
-
-  env <- parent.frame(2)
+  if (identical(fn[1], "get")) {
+    if (length(fn) == 1) { # TODO: (RK) Actually parse out super::multi_parent_env
+      stop("super::super does not support ", sQuote("get"), " calls ",
+           "due to internal purposes.")
+    } else {
+      # From the magic get() call below, we extract the function name.
+      fn <- fn[2] 
+      env <- parent.frame()
+      level <- 2 # See comment below.
+    }
+  } else {
+    env   <- parent.frame(2)
+    level <- 7
+  }
 
   first <- TRUE
+  parent_count <- -1
   while (!identical(env, emptyenv())) {
+    parent_count <- parent_count + 1
     if (exists(fn, envir = env, inherits = FALSE)) {
       super_fn <- get(fn, envir = env, inherits = FALSE)
       if (is.function(super_fn)) {
         if (first) { first <- FALSE }
-        else { return(do.call(fn, eval(substitute(alist(...))), envir = env)) }
+        else {
+          # We have to find the equivalent of fn in the environment that is the
+          # n-fold parent environment of the calling environment, parent.frame(),
+          # where n = parent_count. Otherwise, non-standard evaluation will not work
+          # correctly with super.
+          return(eval.parent(substitute(
+            # The magic number level = 7 comes from the fact eval.parent(...) creates its own
+            # call chain with 5 steps inserted that we wish to skip over. Otherwise,
+            # if we are already in a super calling chain, we use level = 2.
+            {
+              `*call*`      <- sys.call(-4)
+              `*call*`[[1]] <- quote(
+                get(fn, envir = super::multi_parent_env(parent.frame(level), parent_count)) # parent.frame(), parent_count))
+              )
+              eval(`*call*`)
+            }
+          )))
+        }
       }
     }
-    env <- parent.env(env)
+    env <- hop(env)
```

Comment:
Instead of `parent.env`, we need `hop` to avoid falling in the gap of namespaces.